[draft] doc/dead_reckoning: What it is and what it does (#139)
heinrich5991/libtw2: doc/dead_reckoning.md

Introduction
============

The server continually sends snapshots and snapshot deltas to all game clients to update them on the game state.
To avoid sending redundant information, snapshots are sent only occasionally and snapshot deltas are used to compactly represent what changed.
As tees move around, their position and other attributes often change on every tick.
Normally, the server would need to include all of these tees in every snapshot delta.

> However, tees have a the system in place called dead reckoning in place to reduce how often a tee update needs to be sent by the server.

Reviewer: ```suggestion
However, tees have a system in place called dead reckoning in place to reduce how often a tee update needs to be sent by the server.
```

The idea of dead reckoning is that the client can run the physics calculations just as well as the server.
So we define a subset of physics we call dead reckoning which the client and server agree upon.
Dead reckoning only considers a singular tee, without interaction with projectiles or other tees.

> Whenever the server sends a tee update, the state includes the tick on which from the state is.

Reviewer: ```suggestion
Whenever the server sends a tee update, the state includes the tick that the state is for.
```

The client then runs dead reckoning on the single tee.
The server knows which physics are run by the client and only sends a new tee state if the real tee position diverged from the dead reckoning simulation.
Alternatively, if the tee state the client has is 3 seconds old (150 ticks), the server also sends an update.

Assumptions
===========

- Snapshot ticks are strictly monotonically increasing
- Tee ticks are always equal or lower than the snapshot tick
- Tee ticks may not be monotonically increasing!
- The same tee's snap id along with the same tick always identify the same tee state(?)
- The maximum difference between the snapshot tick and tee tick is 150 ticks

Prediction
==========

If the client receives a snapshot with tick `n`, and finds a tee with tick `t`, it runs dead reckoning for `n - t` ticks to get the up-to-date tee.

> In case the client has cached a dead reckoned tee from the original tick `t`, it can also use it to perform less dead reckoning.

Reviewer: You could also use LaTeX math notation such as `$n - t$` to get $n - t$ instead of using inline code.

Reviewer 2: Something is broken about github math handling:

<img width="1404" height="186" src="https://github.com/user-attachments/assets/fb3edc7a-db84-4bf9-8178-e82e13c1d77e" />

A future snapshot does not have a "more correct" tee for this tick, even if that snapshot's tick is further back in the past.

Demos

> -----

Reviewer: Is this supposed to be a different headline size?

Author: Was intended, can ofc be changed

Reviewer: It doesn't really look like it should be a subsection of the `Prediction` section. It also doesn't really apply to demos I think.

> In demos, we receive a snapshot only every second tick and some snapshots are even fully left out.

Servers send a snapshot only every second tick by default. With `sv_high_bandwidth 1` servers send a snapshot every tick. Demos always contain the snapshots sent by the server at the rate sent by the server.

Author: Thanks for the insight, tried to address that :)
My tired brain didn't really come up with a better subsection / subsection name. It is definitely content I want/need clarified.

As tees move around, their position and other attributes often change on every tick.
Normally, the server would need to include all of these tees in every snapshot delta.
However, tees have a system in place called dead reckoning in place to reduce how often a tee update needs to be sent by the server.

The idea of dead reckoning is that the client can run the physics calculations just as well as the server.
Dead reckoning is a subset of Teeworlds physics which the client and server are aware of.
Whenever the server sends a tee update, the state includes the tick that the state is for.
The server knows which physics are run by the client and only sends a new tee state if the real tee position diverged from the dead reckoning simulation.
Alternatively, if the tee state the client has is 3 seconds old (150 ticks), the server also sends an update.

Physics
=======

Dead reckoning is a subset of Teeworlds physics and doesn't include:

- Projectiles
- Other tees
- Changed tunings
- TODO: what else?

Assumptions
===========

- Snapshot ticks are strictly monotonically increasing
- Tee ticks are always equal or lower than the snapshot tick

> - Tee ticks may not be monotonically increasing!

Reviewer 2: Interesting. What would lead to them not being increasing?

Author: I was highly confused by that as well. Turns out it was a bug in another layer of abstraction.
In [TwSnap](https://gitlab.com/ddnet-rs/twsnap) we have demo reader on top of this one that displays that error, and I already have a good idea what the cause is.

I pushed a commit here with tools/demo_check_reckon, which tells us that the tee ticks are indeed monotonically increasing, apart from a special case, where the tick is suddenly 0. 
I don't know exactly when that happens yet, but it appears to occasionally happen when a tee spawns in.
The tees sometimes start at 0, and they always switch off the 0 right after by the looks of it.

In TwSnap that caused me trouble already, and I decided back then to set the tee's tick to the current snap's tick, if the tee's tick is 0.
Back then I thought that was a solid fix, but apparently I didn't test it properly.
That would definitely explain why the tick appears to decrement afterwards when the server sends a proper tick for the tee.

- The same tee's snap id along with the same tick always identify the same tee state(?)
- The maximum difference between the snapshot tick and tee tick is 150 ticks

Prediction
==========

If the client receives a snapshot with tick $n$, and finds a tee with tick $t$, it runs dead reckoning for $n - t$ ticks to get the up-to-date tee.
In case the client has cached a dead reckoned tee from the original tick $t$, it can also use it to perform less dead reckoning.
A future snapshot does not have a "more correct" tee for this tick, even if that snapshot's tick is further back in the past.

Handling Missing Snapshots
==========================

The server can decide to send fewer than one snapshot per tick.
With `sv_high_bandwidth 0`, the server will send at most one snapshot every other tick.

> Longer snapshot gaps mean that dead reckoning produces the expected snapshot, which was able to be left out for that reason(?).

Reviewer 2: Does dead reckoning enter the picture here at all?

Author: I don't know. I was hoping someone else could chime in here ^^

Author: A plausible interpretation would be that dead reckoning is only performed to get a tee's values up-to-date to the snapshot's tick. From there, other prediction could set in to fill the gaps between two snaps. E.g. very simple prediction using the `CharacterCore`s `velocity` field.
